src/main/java/edu/harvard/iq/dataverse/api/MxrdrWorkflowApi.java

```diff
@@ -0,0 +1,125 @@
+        try {
+            AuthenticatedUser user = findAuthenticatedUserOrDie();
+            if (!user.isSuperuser()) {
+                return error(Response.Status.FORBIDDEN, "Only superusers can run workflow");
+            }
+            if (type == null) {
+                return error(Response.Status.BAD_REQUEST, "Missing processing type - 'type=failedOnly' or 'type=notPerformedOnly' is required");
+            }
+
+            int datasetProcessed = 0;
+            for (Dataset dataset:datasetRepository.findAll()) {
+                DatasetVersion released = dataset.getReleasedVersion();
+                if (released != null) {
+                    Optional<WorkflowExecution> result = workflowExecutionService.findLatestByTriggerTypeAndDatasetVersion(TriggerType.PostPublishDataset, 
+                            dataset.getId(), released.getVersionNumber(), released.getMinorVersionNumber());
+                    if (result.isPresent() && type.equals("failedOnly")) {
+                        WorkflowExecution execution = result.get();
+                        if (execution.isFinished() && !execution.getLastStep().getFinishedSuccessfully()) {
+                            rerun(dataset, released);
+                            datasetProcessed++;
+                        }
+                    } else if (!result.isPresent() && type.equals("notPerformedOnly")) {
+                        rerun(dataset, released);
+                        datasetProcessed++;
+                    }
+                }
+            }
+            
+            return ok("Processed " + datasetProcessed + " datasets");
+        } catch (WrappedResponse ex) {
+            return ex.getResponse();
+        } catch (NoDatasetFilesException ex) {
+            return error(Response.Status.INTERNAL_SERVER_ERROR,
+                         "Unable to publish dataset, since there are no files in it.");
+        }
+
+    }
+
+    @POST
+    @Path("{id}/workflow/rerun")
+    public Response rerunWorkflow(@PathParam("id") String id, @QueryParam("version") String version) {
+        try {
+            AuthenticatedUser user = findAuthenticatedUserOrDie();
+            if (!user.isSuperuser()) {
+                return error(Response.Status.FORBIDDEN, "Only superusers can run workflow");
+            }
+            Dataset ds = findDatasetOrDie(id);
+            DatasetVersion updateVersion = null;
+            if (version != null) {
+                for (DatasetVersion datasetVersion:ds.getVersions()) {
+                    if (version.equals(datasetVersion.getVersionNumber() + "." + datasetVersion.getMinorVersionNumber())) {
+                        updateVersion = datasetVersion;
+                    }
+                }
+                if (updateVersion == null) {
+                    return error(Response.Status.BAD_REQUEST, "Unknown version: " + version);
+                } else {
+                    rerun(ds, updateVersion);
```

Comment:
Correct me if I'm wrong but there is a possibility that we potentially end up with workflow in progress and dataset would not have any lock.
for example if we run this endpoint two times for the same dataset:
Run workflow nr 1:
1) add workflow lock
2) add execution to queue
Run workflow nr 2:
1) add workflow lock
2) add execution to queue
Processing:
1) workflow nr 1 finishes
2) workflow nr 1 removes locks
... dataset is unlocked ...
3) workflow nr 2 is executing